File: src/com/secretary/helloworld/ui/dialogs/CompletionDialog.java
package com.secretary.helloworld.ui.dialogs;

import android.app.AlertDialog;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.*;

import com.secretary.helloworld.R;
import com.secretary.helloworld.Task;
import com.secretary.helloworld.AppLogger;
import com.secretary.helloworld.data.dao.TaskDao;
import com.secretary.helloworld.data.dao.CompletionDao;
import com.secretary.helloworld.utils.StatsManager;

/**
 * Dialog for completing a task with additional tracking information.
 * Extracted from TaskActivity for better code organization.
 */
public class CompletionDialog {
    private static final String TAG = "CompletionDialog";

    private final Context context;
    private final TaskDao taskDao;
    private final CompletionDao completionDao;
    private final StatsManager statsManager;
    private final AppLogger logger;

    private AlertDialog dialog;
    private CompletionListener listener;

    /**
     * Listener interface for completion events
     */
    public interface CompletionListener {
        void onTaskCompleted(Task task);
        void onCompletionCancelled(Task task);
    }

    public CompletionDialog(Context context, CompletionListener listener) {
        this.context = context;
        this.taskDao = new TaskDao(context);
        this.completionDao = new CompletionDao(context);
        this.statsManager = new StatsManager(context);
        this.logger = AppLogger.getInstance(context);
        this.listener = listener;
    }

    /**
     * Show the completion dialog for a task
     */
    public void show(Task task) {
        AlertDialog.Builder builder = new AlertDialog.Builder(context);
        View dialogView = LayoutInflater.from(context).inflate(R.layout.dialog_completion, null);
        builder.setView(dialogView);

        dialog = builder.create();

        // Find dialog views
        TextView taskTitleText = dialogView.findViewById(R.id.completionTaskTitle);
        EditText timeSpentInput = dialogView.findViewById(R.id.timeSpentInput);
        TextView averageTimeText = dialogView.findViewById(R.id.averageTimeText);
        SeekBar difficultySeekBar = dialogView.findViewById(R.id.difficultySeekBar);
        TextView difficultyValueText = dialogView.findViewById(R.id.difficultyValueText);
        EditText notesInput = dialogView.findViewById(R.id.completionNotesInput);
        CheckBox quickCompleteCheckBox = dialogView.findViewById(R.id.quickCompleteCheckBox);
        Button skipButton = dialogView.findViewById(R.id.skipDetailsButton);
        Button saveButton = dialogView.findViewById(R.id.saveCompletionButton);

        // Set task title
        taskTitleText.setText(task.getTitle());

        // Show average time if available
        int avgTime = completionDao.getAverageCompletionTime(task.getId());
        if (avgTime > 0) {
            averageTimeText.setText("(Avg: " + avgTime + " min)");
            timeSpentInput.setText(String.valueOf(avgTime));
        }

        // Setup difficulty slider
        difficultySeekBar.setMax(10);
        difficultySeekBar.setProgress(5);
        difficultyValueText.setText("5");

        difficultySeekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                difficultyValueText.setText(String.valueOf(progress));
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {}

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {}
        });

        // Quick complete checkbox
        quickCompleteCheckBox.setOnCheckedChangeListener((buttonView, isChecked) -> {
            int visibility = isChecked ? View.GONE : View.VISIBLE;
            timeSpentInput.setVisibility(visibility);
            averageTimeText.setVisibility(visibility);
            difficultySeekBar.setVisibility(visibility);
            difficultyValueText.setVisibility(visibility);
            notesInput.setVisibility(visibility);

            // Find the parent layouts and hide/show them
            View timeLayout = dialogView.findViewById(R.id.timeSpentInput).getParent();
            View difficultyLayout = dialogView.findViewById(R.id.difficultySeekBar).getParent();
            View notesLayout = dialogView.findViewById(R.id.completionNotesInput).getParent();

            if (timeLayout instanceof View) {
                ((View) timeLayout).setVisibility(visibility);
            }
            if (difficultyLayout instanceof View) {
                ((View) difficultyLayout).setVisibility(visibility);
            }
            if (notesLayout instanceof View) {
                ((View) notesLayout).setVisibility(visibility);
            }
        });

        // Skip button - complete without tracking
        skipButton.setOnClickListener(v -> {
            completeTaskBasic(task);
            dialog.dismiss();
        });

        // Save button - complete with tracking
        saveButton.setOnClickListener(v -> {
            if (quickCompleteCheckBox.isChecked()) {
                // Quick complete
                completeTaskBasic(task);
            } else {
                // Complete with tracking
                String timeStr = timeSpentInput.getText().toString().trim();
                int timeSpent = 0;
                try {
                    timeSpent = Integer.parseInt(timeStr);
                } catch (NumberFormatException e) {
                    // Keep as 0
                }

                int difficulty = difficultySeekBar.getProgress();
                String notes = notesInput.getText().toString().trim();

                completeTaskWithTracking(task, timeSpent, difficulty, notes);
            }
            dialog.dismiss();
        });

        // Handle dialog cancel
        dialog.setOnCancelListener(dialogInterface -> {
            if (listener != null) {
                listener.onCompletionCancelled(task);
            }
        });

        dialog.show();
    }

    /**
     * Complete task without tracking
     */
    private void completeTaskBasic(Task task) {
        task.setCompleted(true);
        task.setLastCompletedDate(System.currentTimeMillis());

        // Handle recurrence
        if (task.isRecurring()) {
            handleRecurringTaskCompletion(task);
        }

        // Update streak
        statsManager.updateTaskStreak(task);

        // Save to database
        taskDao.updateTask(task);
        taskDao.markTaskCompleted(task.getId(), true);

        logger.info(TAG, "Task completed (basic): " + task.getTitle());

        if (listener != null) {
            listener.onTaskCompleted(task);
        }
    }

    /**
     * Complete task with detailed tracking
     */
    private void completeTaskWithTracking(Task task, int timeSpent, int difficulty, String notes) {
        // Record completion details
        completionDao.recordCompletion(task.getId(), timeSpent, difficulty, notes);

        // Mark task as completed
        task.setCompleted(true);
        task.setLastCompletedDate(System.currentTimeMillis());

        // Handle recurrence
        if (task.isRecurring()) {
            handleRecurringTaskCompletion(task);
        }

        // Update streak
        statsManager.updateTaskStreak(task);

        // Save to database
        taskDao.updateTask(task);
        taskDao.markTaskCompleted(task.getId(), true);

        logger.info(TAG, "Task completed with tracking: " + task.getTitle() +
                    " (Time: " + timeSpent + " min, Difficulty: " + difficulty + ")");

        if (listener != null) {
            listener.onTaskCompleted(task);
        }
    }

    /**
     * Handle completion for recurring tasks
     */
    private void handleRecurringTaskCompletion(Task task) {
        if (task.getRecurrenceType() == Task.RECURRENCE_FREQUENCY) {
            // Update frequency count
            int completions = task.getCompletionsThisPeriod() + 1;
            task.setCompletionsThisPeriod(completions);

            // Check if period needs reset
            long periodStart = task.getCurrentPeriodStart();
            long periodEnd = calculatePeriodEnd(periodStart, task.getRecurrenceUnit());

            if (System.currentTimeMillis() > periodEnd) {
                // Start new period
                task.setCurrentPeriodStart(System.currentTimeMillis());
                task.setCompletionsThisPeriod(1);
            } else {
                task.setCompletionsThisPeriod(completions);
            }

            // Check if goal is met
            if (completions >= task.getRecurrenceAmount()) {
                // Goal met, task stays completed
                task.setCompleted(true);
            } else {
                // More completions needed
                task.setCompleted(false);
            }
        }
        // For interval tasks, they stay completed and will reappear after the interval
    }

    /**
     * Calculate when a period ends based on unit
     */
    private long calculatePeriodEnd(long periodStart, int unit) {
        long duration;
        switch (unit) {
            case Task.UNIT_WEEK:
                duration = 7L * 24 * 60 * 60 * 1000;
                break;
            case Task.UNIT_MONTH:
                duration = 30L * 24 * 60 * 60 * 1000;
                break;
            default: // UNIT_DAY
                duration = 24L * 60 * 60 * 1000;
                break;
        }
        return periodStart + duration;
    }

    /**
     * Dismiss the dialog
     */
    public void dismiss() {
        if (dialog != null && dialog.isShowing()) {
            dialog.dismiss();
        }
    }
}

File: app/src/main/java/com/secretary/features/tasks/presentation/dialog/CompletionDialog.kt
package com.secretary.features.tasks.presentation.dialog

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.*
import androidx.core.os.bundleOf
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.setFragmentResult
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import com.secretary.R
import com.secretary.features.tasks.domain.model.Task
import com.secretary.features.statistics.data.CompletionRepositoryImpl
import com.secretary.features.statistics.domain.repository.CompletionRepository
import com.secretary.features.tasks.presentation.viewmodel.TaskListViewModel
import com.secretary.features.tasks.presentation.viewmodel.TaskViewModelFactory
import com.secretary.features.dreams.domain.usecase.CompleteTaskWithXPUseCase
import com.secretary.features.dreams.data.repository.TaskMilestoneLinkRepositoryImpl
import com.secretary.features.dreams.data.repository.MilestoneRepositoryImpl
import com.secretary.features.dreams.data.repository.DreamRepositoryImpl
import com.secretary.features.dreams.domain.service.XPCalculationService
import com.secretary.features.dreams.domain.service.LevelProgressionService
import kotlinx.coroutines.launch

/**
 * DialogFragment for completing a task with optional tracking
 * Phase 4.5.6: Dialog Extraction
 *
 * Replaces TaskDialogHelper.showCompletionDialog()
 * Uses TaskListViewModel for task completion + CompletionRepository for tracking
 */
class CompletionDialog : DialogFragment() {

    private lateinit var viewModel: TaskListViewModel
    private lateinit var completionRepository: CompletionRepository
    private var task: Task? = null

    companion object {
        const val TAG = "CompletionDialog"
        const val RESULT_KEY = "task_completed"
        const val EXTRA_TASK_ID = "task_id"
        const val EXTRA_TASK_TITLE = "task_title"
        const val EXTRA_TASK_PRIORITY = "task_priority"

        fun newInstance(task: Task, factory: TaskViewModelFactory): CompletionDialog {
            return CompletionDialog().apply {
                arguments = bundleOf(
                    EXTRA_TASK_ID to task.id,
                    EXTRA_TASK_TITLE to task.title,
                    EXTRA_TASK_PRIORITY to task.priority
                )
            }
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val taskId = arguments?.getLong(EXTRA_TASK_ID) ?: 0
        val taskTitle = arguments?.getString(EXTRA_TASK_TITLE) ?: "Task"
        val taskPriority = arguments?.getInt(EXTRA_TASK_PRIORITY) ?: 1

        // Create a dummy task object (we only need id, title, and priority)
        task = Task().apply {
            this.id = taskId
            this.title = taskTitle
            this.priority = taskPriority
        }

        val dialogView = LayoutInflater.from(requireContext()).inflate(R.layout.dialog_completion, null)
        val dialog = androidx.appcompat.app.AlertDialog.Builder(requireContext())
            .setView(dialogView)
            .create()

        setupViewModel()
        setupViews(dialogView)

        // Load average completion time
        loadAverageTime(dialogView, taskId)

        // Load XP preview
        loadXPPreview(dialogView, taskId, task?.priority ?: 1)

        return dialog
    }

    private fun setupViewModel() {
        // Setup TaskListViewModel for task completion
        val database = com.secretary.shared.database.TaskDatabase.getDatabase(requireContext())
        val taskDao = database.taskDao()
        val taskRepository = com.secretary.features.tasks.data.repository.TaskRepositoryImpl(taskDao)
        val streakService = com.secretary.features.tasks.domain.service.StreakService()
        val recurrenceService = com.secretary.features.tasks.domain.service.RecurrenceService()
        val motivationalMessageService = com.secretary.features.motivation.domain.MotivationalMessageService()

        // Setup CompletionRepository for tracking data
        val completionDao = database.completionDao()
        completionRepository = CompletionRepositoryImpl(completionDao)

        val factory = TaskViewModelFactory(taskRepository, completionRepository, streakService, recurrenceService, motivationalMessageService)
        viewModel = ViewModelProvider(this, factory)[TaskListViewModel::class.java]

        // Observe completion result
        viewModel.operationSuccess.observe(this) { message ->
            message?.let {
                Toast.makeText(requireContext(), it, Toast.LENGTH_SHORT).show()
                setFragmentResult(RESULT_KEY, bundleOf("success" to true))
                dismiss()
            }
        }

        // Observe errors
        viewModel.error.observe(this) { error ->
            error?.let {
                Toast.makeText(requireContext(), it, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun setupViews(dialogView: View) {
        // Find views
        val taskTitleText = dialogView.findViewById<TextView>(R.id.completionTaskTitle)
        val timeSpentInput = dialogView.findViewById<EditText>(R.id.timeSpentInput)
        val averageTimeText = dialogView.findViewById<TextView>(R.id.averageTimeText)
        val difficultySeekBar = dialogView.findViewById<SeekBar>(R.id.difficultySeekBar)
        val difficultyValueText = dialogView.findViewById<TextView>(R.id.difficultyValueText)
        val notesInput = dialogView.findViewById<EditText>(R.id.completionNotesInput)
        val quickCompleteCheckBox = dialogView.findViewById<CheckBox>(R.id.quickCompleteCheckBox)
        val skipButton = dialogView.findViewById<Button>(R.id.skipDetailsButton)
        val saveButton = dialogView.findViewById<Button>(R.id.saveCompletionButton)

        // Set task title
        taskTitleText.text = task?.title ?: "Task"

        // Setup difficulty seekbar
        difficultySeekBar.max = 10
        difficultySeekBar.progress = 5 // Default to medium
        difficultyValueText.text = "5"

        difficultySeekBar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                difficultyValueText.text = progress.toString()
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {}
        })

        // Quick complete checkbox
        quickCompleteCheckBox.setOnCheckedChangeListener { _, isChecked ->
            val visibility = if (isChecked) View.GONE else View.VISIBLE
            timeSpentInput.visibility = visibility
            averageTimeText.visibility = visibility
            difficultySeekBar.visibility = visibility
            difficultyValueText.visibility = visibility
            notesInput.visibility = visibility
        }

        // Skip button - complete without tracking
        skipButton.setOnClickListener {
            completeTaskBasic()
        }

        // Save button - complete with tracking
        saveButton.setOnClickListener {
            if (quickCompleteCheckBox.isChecked) {
                completeTaskBasic()
            } else {
                val timeSpent = try {
                    timeSpentInput.text.toString().toInt()
                } catch (e: NumberFormatException) {
                    0
                }

                val difficulty = difficultySeekBar.progress
                val notes = notesInput.text.toString().trim()

                completeTaskWithTracking(timeSpent, difficulty, notes)
            }
        }

        // Handle dialog cancel
        dialog?.setOnCancelListener {
            setFragmentResult(RESULT_KEY, bundleOf("cancelled" to true))
        }
    }

    private fun loadAverageTime(dialogView: View, taskId: Long) {
        val averageTimeText = dialogView.findViewById<TextView>(R.id.averageTimeText)

        lifecycleScope.launch {
            try {
                val avgTime = completionRepository.getAverageCompletionTime(taskId)
                if (avgTime > 0) {
                    averageTimeText.text = "(Avg: $avgTime min)"
                    averageTimeText.visibility = View.VISIBLE
                } else {
                    averageTimeText.visibility = View.GONE
                }
            } catch (e: Exception) {
                // Failed to load average time - not critical, just hide
                averageTimeText.visibility = View.GONE
            }
        }
    }

    /**
     * Load XP preview for the task.
     * Shows how much XP will be earned from completing this task.
     */
    private fun loadXPPreview(dialogView: View, taskId: Long, taskPriority: Int) {
        val xpPreviewSection = dialogView.findViewById<LinearLayout>(R.id.xpPreviewSection)
        val xpPreviewText = dialogView.findViewById<TextView>(R.id.xpPreviewText)
        val milestonePreviewText = dialogView.findViewById<TextView>(R.id.milestonePreviewText)

        lifecycleScope.launch {
            try {
                val database = com.secretary.shared.database.TaskDatabase.getDatabase(requireContext())

                // Create repositories
                val linkRepo = TaskMilestoneLinkRepositoryImpl(database.taskMilestoneJunctionDao())
                val milestoneRepo = MilestoneRepositoryImpl(database.milestoneDao())
                val dreamRepo = DreamRepositoryImpl(database.dreamDao())

                // Create services
                val xpService = XPCalculationService()
                val levelService = LevelProgressionService()

                // Create use case
                val useCase = CompleteTaskWithXPUseCase(
                    linkRepo, milestoneRepo, dreamRepo, xpService, levelService
                )

                // Get preview
                val preview = useCase.previewXP(taskId, taskPriority)

                // Update UI on main thread
                activity?.runOnUiThread {
                    if (preview.milestoneCount > 0) {
                        xpPreviewSection.visibility = View.VISIBLE

                        // Format XP preview text
                        if (preview.milestoneCount == 1) {
                            xpPreviewText.text = "+${preview.xpPerMilestone} XP"
                        } else {
                            xpPreviewText.text = "+${preview.xpPerMilestone} XP × ${preview.milestoneCount} = ${preview.totalXp} XP total"
                        }

                        // Format milestone list
                        if (preview.linkedMilestones.isNotEmpty()) {
                            milestonePreviewText.text = preview.linkedMilestones.joinToString("\n") { "• $it" }
                            milestonePreviewText.visibility = View.VISIBLE
                        } else {
                            milestonePreviewText.visibility = View.GONE
                        }
                    } else {
                        // No milestones linked - hide the section
                        xpPreviewSection.visibility = View.GONE
                    }
                }
            } catch (e: Exception) {
                // Failed to load XP preview - hide section
                // This is not critical, so we just silently hide it
                activity?.runOnUiThread {
                    xpPreviewSection?.visibility = View.GONE
                }
            }
        }
    }

    /**
     * Complete task without detailed tracking
     */
    private fun completeTaskBasic() {
        val taskId = task?.id ?: return

        lifecycleScope.launch {
            viewModel.completeTask(taskId)
        }
    }

    /**
     * Complete task with time and difficulty tracking
     */
    private fun completeTaskWithTracking(timeSpent: Int, difficulty: Int, notes: String) {
        val taskId = task?.id ?: return

        lifecycleScope.launch {
            try {
                // Save completion tracking data first
                completionRepository.saveCompletion(
                    taskId = taskId,
                    timeSpentMinutes = timeSpent,
                    difficulty = difficulty,
                    notes = notes.ifEmpty { null }
                )

                // Then complete the task (updates streaks, handles recurrence)
                viewModel.completeTask(taskId)
            } catch (e: Exception) {
                Toast.makeText(
                    requireContext(),
                    "Failed to save completion data: ${e.message}",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
    }
}
